pipeline/pipeline.py

```diff
@@ -0,0 +1,53 @@
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import List, Optional
+
+from .task import AbstractTask
+
+
+class Pipeline:
+    """Class that coordinates run of a group of tasks, which form a DAG based on dependencies
+    defined in add_task().
+
+    See pipeline_test.py for examples.
+
+    TBD: The pipeline has built-in checkpoint i.e it will skip all complete tasks upon restart."""
+
+    @dataclass
+    class Node:
+        """Node within a pipeline.
+
+        Attributes:
+        - task: The task runs in the node.
+        - children: The nodes that should run after the current node completes.
+        """
+        task: AbstractTask = None
+        children = []
+
+    def __init__(self, working_dir: str):
+        """
+        :param working_dir: the dir where the pipeline reads / writes checkpoints and outputs logs.
+        """
+        self._source = self.Node()
+        self._working_dir = working_dir
+        self._tasks = []
+
+    def add_task(self, task: AbstractTask, parents: Optional[List[Node]] = None) -> Node:
+        node = self.Node(task)
+        if not parents:
+            parents = [self._source]
+        for parent in parents:
```

Comment:
Should we enforce that the passed in parents' tasks indeed exist in self._tasks?

Reply:
Added comments to prevent users creating own Node instances.